### PR TITLE
Add cooldown saving for homunculus and mercenary

### LINKED ISSUE #8518: Homunculus and Mercenary cooldown not saving
### rAthena Hash

e39fa4b66a543645b0f70f394e44a1eccf8a99ee

### Client Date

2018-06-21

### Server Mode

Renewal

### Result

Homunculus and Mercenary cooldowns are reset after relog or vaporize.

### Relevant Log Output

_No response_

### Expected Result

Homunculus and Mercenary cooldowns should be saved in the SQL skillcooldown database and continue after relog and vaporize.

### How to Reproduce

Use any Homunculus or Mercenary skill with a cooldown and relog, after loading the cooldown is gone.

### Official Information

On official servers, those cooldowns are saved and restored after relog.
Elemental skill cooldowns on the other hand are not really cooldowns, but similar to monsters' skill delay.

### Modifications that may affect results

_No response_

### PR DESCRIPTION
**Pull request by aleos89**
* **Addressed Issue(s)**: Fixes #8518

* **Server Mode**: Pre-renewal and Renewal

* **Description of Pull Request**: 
  * Allow skill cooldowns for homunculus and mercenary to be saved when the player logs off.

### CODE REVIEW

File: src/char/char.cpp [hunk 1]
```diff
@@ -2322,8 +2322,8 @@ bool char_checkdb(void){
                 schema_config.guild_expulsion_db, schema_config.guild_member_db, 
                 schema_config.guild_skill_db, schema_config.guild_position_db, schema_config.guild_storage_db,
 		schema_config.party_db, schema_config.pet_db, schema_config.friend_db, schema_config.mail_db, 
-                schema_config.auction_db, schema_config.quest_db, schema_config.homunculus_db, schema_config.skill_homunculus_db,
-                schema_config.mercenary_db, schema_config.mercenary_owner_db,
+                schema_config.auction_db, schema_config.quest_db, schema_config.homunculus_db, schema_config.skill_homunculus_db, schema_config.skillcooldown_homunculus_db,
```
Comment (Lemongrass3110) [inline src/char/char.cpp:2325]:
```suggestion
                schema_config.auction_db, schema_config.quest_db,
                schema_config.homunculus_db, schema_config.skill_homunculus_db, schema_config.skillcooldown_homunculus_db,
```

File: src/char/char_mapif.cpp [hunk 2]
```diff
@@ -517,15 +517,15 @@ int chmapif_parse_req_saveskillcooldown(int fd){
 		count = RFIFOW(fd,12);
 		if( count > 0 )
 		{
-			struct skill_cooldown_data data;
+			struct s_skill_cooldown_data data;
```
Comment (Lemongrass3110) [inline src/char/char_mapif.cpp:520]:
```suggestion
			s_skill_cooldown_data data;
```

File: src/char/char_mapif.cpp [hunk 3]
```diff
@@ -517,15 +517,15 @@ int chmapif_parse_req_saveskillcooldown(int fd){
...
 			StringBuf buf;
 			int i;
 
 			StringBuf_Init(&buf);
 			StringBuf_Printf(&buf, "INSERT INTO `%s` (`account_id`, `char_id`, `skill`, `tick`) VALUES ", schema_config.skillcooldown_db);
 			for( i = 0; i < count; ++i )
 			{
-				memcpy(&data,RFIFOP(fd,14+i*sizeof(struct skill_cooldown_data)),sizeof(struct skill_cooldown_data));
+				memcpy(&data,RFIFOP(fd,14+i*sizeof(struct s_skill_cooldown_data)),sizeof(struct s_skill_cooldown_data));
```
Comment (Lemongrass3110) [inline src/char/char_mapif.cpp:528]:
```suggestion
				memcpy(&data,RFIFOP(fd,14+i*sizeof(s_skill_cooldown_data)),sizeof(s_skill_cooldown_data));
```

File: src/char/char_mapif.cpp [hunk 4]
```diff
@@ -558,23 +558,23 @@ int chmapif_parse_req_skillcooldown(int fd){
 		{
 			int count;
 			char* data;
-			struct skill_cooldown_data scd;
+			struct s_skill_cooldown_data scd;
```
Comment (Lemongrass3110) [inline src/char/char_mapif.cpp:561]:
```suggestion
			s_skill_cooldown_data scd;
```

File: src/char/char_mapif.cpp [hunk 5]
```diff
@@ -558,23 +558,23 @@ int chmapif_parse_req_skillcooldown(int fd){
...
 
-			WFIFOHEAD(fd,14 + MAX_SKILLCOOLDOWN * sizeof(struct skill_cooldown_data));
+			WFIFOHEAD(fd,14 + MAX_SKILLCOOLDOWN * sizeof(struct s_skill_cooldown_data));
```
Comment (Lemongrass3110) [inline src/char/char_mapif.cpp:563]:
```suggestion
			WFIFOHEAD(fd,14 + MAX_SKILLCOOLDOWN * sizeof(s_skill_cooldown_data));
```

File: src/char/char_mapif.cpp [hunk 6]
```diff
@@ -558,23 +558,23 @@ int chmapif_parse_req_skillcooldown(int fd){
...
 			WFIFOW(fd,0) = 0x2b0b;
 			WFIFOL(fd,4) = aid;
 			WFIFOL(fd,8) = cid;
 			for( count = 0; count < MAX_SKILLCOOLDOWN && SQL_SUCCESS == Sql_NextRow(sql_handle); ++count )
 			{
 				Sql_GetData(sql_handle, 0, &data, nullptr); scd.skill_id = atoi(data);
 				Sql_GetData(sql_handle, 1, &data, nullptr); scd.tick = strtoll( data, nullptr, 10 );
-				memcpy(WFIFOP(fd,14+count*sizeof(struct skill_cooldown_data)), &scd, sizeof(struct skill_cooldown_data));
+				memcpy(WFIFOP(fd,14+count*sizeof(struct s_skill_cooldown_data)), &scd, sizeof(struct s_skill_cooldown_data));
```
Comment (Lemongrass3110) [inline src/char/char_mapif.cpp:571]:
```suggestion
				memcpy(WFIFOP(fd,14+count*sizeof(s_skill_cooldown_data)), &scd, sizeof(s_skill_cooldown_data));
```

File: src/char/char_mapif.cpp [hunk 7]
```diff
@@ -558,23 +558,23 @@ int chmapif_parse_req_skillcooldown(int fd){
...
 			}
 			if( count >= MAX_SKILLCOOLDOWN )
 				ShowWarning("Too many skillcooldowns for %d:%d, some of them were not loaded.\n", aid, cid);
 			if( count > 0 )
 			{
-				WFIFOW( fd, 2 ) = static_cast<int16>( 14 + count * sizeof( struct skill_cooldown_data ) );
+				WFIFOW( fd, 2 ) = static_cast<int16>( 14 + count * sizeof( struct s_skill_cooldown_data ) );
```
Comment (Lemongrass3110) [inline src/char/char_mapif.cpp:577]:
```suggestion
				WFIFOW( fd, 2 ) = static_cast<int16>( 14 + count * sizeof( s_skill_cooldown_data ) );
```

File: src/char/int_homun.cpp [hunk 8]
```diff
@@ -142,6 +141,25 @@ bool mapif_homunculus_save(struct s_homunculus* hd)
 		}
 	}
 
+	// Save skill cooldowns
+	SqlStmt *stmt = SqlStmt_Malloc(sql_handle);
+
+	if (SQL_ERROR == SqlStmt_Prepare(stmt, "REPLACE INTO `%s` (`homun_id`, `skill`, `tick`) VALUES (%d, ?, ?)", schema_config.skillcooldown_homunculus_db, hd->hom_id))
+		SqlStmt_ShowDebug(stmt);
```
Comment (Lemongrass3110) [inline src/char/int_homun.cpp:148]: If it fails to prepare, it does not really make sense to continue?
But I see that this is already wrong in the code where you copied it from.

File: src/char/int_homun.cpp [hunk 9]
```diff
@@ -142,6 +141,25 @@ bool mapif_homunculus_save(struct s_homunculus* hd)
...
+	for (uint16 i = 0; i < MAX_SKILLCOOLDOWN; ++i) {
+		if (hd->scd[i].skill_id > 0 && hd->scd[i].tick != 0) {
```
Comment (Lemongrass3110) [inline src/char/int_homun.cpp:150]:
```suggestion
		if (hd->scd[i].skill_id == 0) {
			continue;
		}

		if (hd->scd[i].tick == 0) {
			continue;
		}
```

File: src/char/int_homun.cpp [hunk 10]
```diff
@@ -142,6 +141,25 @@ bool mapif_homunculus_save(struct s_homunculus* hd)
...
+			SqlStmt_BindParam(stmt, 0, SQLDT_USHORT, &hd->scd[i].skill_id, 0);
+			SqlStmt_BindParam(stmt, 1, SQLDT_LONGLONG, &hd->scd[i].tick, 0);
+			if (SQL_ERROR == SqlStmt_Execute(stmt)) {
+				SqlStmt_ShowDebug(stmt);
+				SqlStmt_Free(stmt);
+				flag = false;
+				break;
+			}
+		}
```
Comment (Lemongrass3110) [inline src/char/int_homun.cpp:159]:
```suggestion
		if ( SQL_ERROR  == SqlStmt_BindParam(stmt, 0, SQLDT_USHORT, &hd->scd[i].skill_id, 0)
		|| SQL_ERROR  == SqlStmt_BindParam(stmt, 1, SQLDT_LONGLONG, &hd->scd[i].tick, 0)
		|| SQL_ERROR == SqlStmt_Execute(stmt)) {
			SqlStmt_ShowDebug(stmt);
			flag = false;
			break;
		}
```
Issues fixed:
Missing SQL_ERROR checks for the parameter binding
Double free of prepared statement.

File: src/char/int_homun.cpp [hunk 11]
```diff
@@ -225,6 +243,42 @@ bool mapif_homunculus_load(int homun_id, struct s_homunculus* hd)
...
+
+	while (SQL_SUCCESS == Sql_NextRow(sql_handle)) {
+		if (count == MAX_SKILLCOOLDOWN) {
+			ShowWarning("Too many skillcooldowns for homunculus %d, skipping.\n", homun_id);
+			break;
+		}
+
+		// Skill
+		Sql_GetData(sql_handle, 0, &data, nullptr);
+		uint16 skill_id = static_cast<uint16>(atoi(data));
```
Comment (Lemongrass3110) [inline src/char/int_homun.cpp:262]:
```suggestion
		uint16 skill_id = static_cast<uint16>(strtoul(data, nullptr, 10));
```

File: src/char/int_mercenary.cpp [hunk 12]
```diff
@@ -57,6 +57,24 @@ bool mercenary_owner_tosql(uint32 char_id, struct mmo_charstatus *status)
...
+		Sql_FreeResult(sql_handle);
+		return false;
+	}
+
+	char *data;
+	int32 mer_id;
+
+	Sql_GetData(sql_handle, 0, &data, nullptr); mer_id = atoi(data);
+
+	if (SQL_ERROR == Sql_Query(sql_handle, "DELETE FROM `%s` WHERE `mer_id` = '%d'", schema_config.skillcooldown_mercenary_db, mer_id))
```
Comment (Lemongrass3110) [inline src/char/int_mercenary.cpp:75]:
```suggestion
	if (SQL_ERROR == Sql_Query(sql_handle, "DELETE FROM `%s` WHERE `mer_id` IN ( SELECT `merc_id` FROM `%s` WHERE `char_id` = '%d' )", schema_config.skillcooldown_mercenary_db, schema_config.mercenary_owner_db, char_id))
```

File: src/char/int_mercenary.cpp [hunk 13]
```diff
@@ -57,6 +57,24 @@ bool mercenary_owner_tosql(uint32 char_id, struct mmo_charstatus *status)
...
+	if (SQL_SUCCESS != Sql_NextRow(sql_handle)) {
+		Sql_FreeResult(sql_handle);
+		return false;
+	}
+
+	char *data;
+	int32 mer_id;
+
+	Sql_GetData(sql_handle, 0, &data, nullptr); mer_id = atoi(data);
+
```
Comment (Lemongrass3110) [inline src/char/int_mercenary.cpp:74]:
```suggestion
```

File: src/char/int_mercenary.cpp [hunk 14]
```diff
@@ -90,6 +108,25 @@ bool mapif_mercenary_save(struct s_mercenary* merc)
 		flag = false;
 	}
 
+	// Save skill cooldowns
+	SqlStmt *stmt = SqlStmt_Malloc(sql_handle);
+
+	if (SQL_ERROR == SqlStmt_Prepare(stmt, "REPLACE INTO `%s` (`mer_id`, `skill`, `tick`) VALUES (%d, ?, ?)", schema_config.skillcooldown_mercenary_db, merc->mercenary_id))
+		SqlStmt_ShowDebug(stmt);
```
Comment (Lemongrass3110) [inline src/char/int_mercenary.cpp:115]: If it fails to prepare, it does not really make sense to continue?
But I see that this is already wrong in the code where you copied it from.

File: src/char/int_mercenary.cpp [hunk 15]
```diff
@@ -90,6 +108,25 @@ bool mapif_mercenary_save(struct s_mercenary* merc)
...
+	for (uint16 i = 0; i < MAX_SKILLCOOLDOWN; ++i) {
+		if (merc->scd[i].skill_id > 0 && merc->scd[i].tick != 0) {
```
Comment (Lemongrass3110) [inline src/char/int_mercenary.cpp:117]:
```suggestion
		if (merc->scd[i].skill_id == 0) {
			continue;
		}

		if (merc->scd[i].tick == 0) {
			continue;
		}
```

File: src/char/int_mercenary.cpp [hunk 16]
```diff
@@ -90,6 +108,25 @@ bool mapif_mercenary_save(struct s_mercenary* merc)
...
+			SqlStmt_BindParam(stmt, 0, SQLDT_USHORT, &merc->scd[i].skill_id, 0);
+			SqlStmt_BindParam(stmt, 1, SQLDT_LONGLONG, &merc->scd[i].tick, 0);
+			if (SQL_ERROR == SqlStmt_Execute(stmt)) {
+				SqlStmt_ShowDebug(stmt);
+				SqlStmt_Free(stmt);
+				flag = false;
+				break;
+			}
+		}
```
Comment (Lemongrass3110) [inline src/char/int_mercenary.cpp:126]:
```suggestion
		if ( SQL_ERROR == SqlStmt_BindParam(stmt, 0, SQLDT_USHORT, &merc->scd[i].skill_id, 0)
		|| SQL_ERROR == SqlStmt_BindParam(stmt, 1, SQLDT_LONGLONG, &merc->scd[i].tick, 0)
		|| SQL_ERROR == SqlStmt_Execute(stmt)) {
			SqlStmt_ShowDebug(stmt);
			flag = false;
			break;
		}
```
Issues fixed:
Missing SQL_ERROR checks for the parameter binding
Double free of prepared statement.

File: src/char/int_mercenary.cpp [hunk 17]
```diff
@@ -122,6 +159,42 @@ bool mapif_mercenary_load(int merc_id, uint32 char_id, struct s_mercenary *merc)
...
+
+	while (SQL_SUCCESS == Sql_NextRow(sql_handle)) {
+		if (count == MAX_SKILLCOOLDOWN) {
+			ShowWarning("Too many skillcooldowns for mercenary %d, skipping.\n", merc_id);
+			break;
+		}
+
+		// Skill
+		Sql_GetData(sql_handle, 0, &data, nullptr);
+		uint16 skill_id = static_cast<uint16>(atoi(data));
```
Comment (Lemongrass3110) [inline src/char/int_mercenary.cpp:178]:
```suggestion
		uint16 skill_id = static_cast<uint16>(strtoul(data, nullptr, 10));
```

File: src/map/homunculus.hpp [hunk 18]
```diff
@@ -98,7 +98,7 @@ struct homun_data {
 	map_session_data *master; //pointer back to its master
 	int hungry_timer;	//[orn]
 	t_exp exp_next;
-	std::vector<uint16> blockskill;	// [orn]
+	std::unordered_map<uint16, int32> scd;	// [orn]
```
Comment (Lemongrass3110) [inline src/map/homunculus.hpp:101]:
```suggestion
	std::unordered_map<uint16, int32> scd;
```

File: src/map/homunculus.cpp [hunk 19]
```diff
@@ -816,6 +815,31 @@ void hom_save(struct homun_data *hd)
...
+	uint16 count = 0;
+	t_tick tick = gettick();
+
+	for (const auto &entry : hd->scd) {
+		const TimerData *timer = get_timer(entry.second);
+
+		if (timer == nullptr || timer->func != skill_blockhomun_end || DIFF_TICK(timer->tick, tick) < 0)
+			continue;
+
+		s_skill_cooldown_data data;
```
Comment (Lemongrass3110) [inline src/map/homunculus.cpp:833]:
```suggestion
		s_skill_cooldown_data& data = hd->homunculus.scd[count];
```

File: src/map/homunculus.cpp [hunk 20]
```diff
@@ -816,6 +815,31 @@ void hom_save(struct homun_data *hd)
...
+
+		data.tick = DIFF_TICK(timer->tick, tick);
+		data.skill_id = entry.first;
+
+		hd->homunculus.scd[count] = data;
```
Comment (Lemongrass3110) [inline src/map/homunculus.cpp:838]:
```suggestion
```

File: src/map/mercenary.cpp [hunk 21]
```diff
@@ -237,6 +237,30 @@ void mercenary_save(s_mercenary_data *md) {
...
+	uint16 count = 0;
+	t_tick tick = gettick();
+
+	for (const auto &entry : md->scd) {
+		const TimerData *timer = get_timer(entry.second);
+
+		if (timer == nullptr || timer->func != skill_blockmerc_end || DIFF_TICK(timer->tick, tick) < 0)
+			continue;
+
+		s_skill_cooldown_data data = {};
```
Comment (Lemongrass3110) [inline src/map/mercenary.cpp:254]:
```suggestion
		s_skill_cooldown_data& data = md->mercenary.scd[count];
```

File: src/map/mercenary.cpp [hunk 22]
```diff
@@ -237,6 +237,30 @@ void mercenary_save(s_mercenary_data *md) {
...
+
+		data.tick = DIFF_TICK(timer->tick, tick);
+		data.skill_id = entry.first;
+
+		md->mercenary.scd[count] = data;
```
Comment (Lemongrass3110) [inline src/map/mercenary.cpp:259]:
```suggestion
```

File: src/map/skill.cpp [hunk 23]
```diff
@@ -852,7 +852,7 @@ bool skill_isNotOk( uint16 skill_id, map_session_data& sd ){
 		return true;
 	}
 
-	if (skill_blockpc_get(&sd, skill_id) != -1){
+	if (sd.scd.find(skill_id) != sd.scd.end()) {
```
Comment (Lemongrass3110) [inline src/map/skill.cpp:855]: We should clone map_exists in utilities.hpp as umap_exists for such cases

File: src/map/skill.cpp [hunk 24]
```diff
@@ -13428,20 +13428,31 @@ TIMER_FUNC(skill_castend_id){
...
+				skill_blockhomun_start(hd, ud->skill_id, skill_get_delay(ud->skill_id, ud->skill_lv));
 #endif
-		} break;
+			}
+			break;
+			case BL_MER:
+			{
+				s_mercenary_data &mc = reinterpret_cast<s_mercenary_data &>(*src);
+
+				skill_blockmerc_start(mc, ud->skill_id, skill_get_cooldown(ud->skill_id, ud->skill_lv));
```
Comment (Lemongrass3110) [inline src/map/skill.cpp:13453]: Do we need a ifdef RENEWAL here too?

**Review by Lemongrass3110 — APPROVED** [review]
Looks good! 👍

File: src/map/skill.cpp (hunk 24 above)
Comment (Daegaladh) [reply src/map/skill.cpp:13441]: There is no cooldown at al on pre-re, but we don't have ifdef for player's cooldown on pre-re either

**Review by Lemongrass3110 — APPROVED** [review]

**Pull request merged**